---
int8 num_clusters
bool success